File: smells/commandinjection.js
const vscode = require('vscode');
var smell = {

    detect : (token) => {

        if(token.hasOwnProperty("name")) var name= token.name;
        if(token.hasOwnProperty("line")) var lineno = token.line;
        if(token.hasOwnProperty("type")) var tokenType = token.type;
        if(token.hasOwnProperty("args")) var args = token.args;
        if(token.hasOwnProperty("hasInputs")) var hasInputs= token.hasInputs;
        
        const cmdFuncs = ['sys.argv','ArgumentParser','argparse','subprocess.Popen','os.system'];

        if(tokenType == "function_call" && cmdFuncs.includes(name) && (args.length > 0 || hasInputs == true))
        {
            const warning = 'possible use of command injection at line '+ lineno;
            vscode.window.showWarningMessage(warning);
            // vscode.commands.executeCommand('revealLine',{'lineNumber':lineno, 'at':'top'});
        }
        else if(tokenType == "variable" && token.hasOwnProperty("valueSrc"))
        {
            if(cmdFuncs.includes(token.valueSrc))
            {
                const warning = 'possible use of command injection at line '+ lineno;
                vscode.window.showWarningMessage(warning);
                // vscode.commands.executeCommand('revealLine',{'lineNumber':lineno, 'at':'top'});
            } 
        }
    }
}

module.exports = smell;

File: py/smells/commandinjection.py
from operations.savewarnings import saveWarnings

def detect(token):
    
    if token.__contains__("line"): lineno = token["line"] 
    if token.__contains__("name"): name = token["name"]
    if token.__contains__("type"): tokenType = token["type"]
    if token.__contains__("args"): args = token["args"]
    if token.__contains__("hasInputs"): containsUserInput =  token["hasInputs"]

    unwanted = ['subprocess.Popen']
    
    if tokenType == "function_call" and name in unwanted and (args or containsUserInput):
        warning = 'use of cmd injection'
        saveWarnings(warning,str(lineno))
        print(warning+ ' at line '+ str(lineno))
